### ISSUE TITLE
Decrease cargo test runtimes of test_stash_postgres/test_auth_admin/test_auth_expiry/test_builtin_migration/test_smoketest_all_builtins

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#21891]
### What version of Materialize are you using?

main

### What is the issue?

As noted in https://github.com/MaterializeInc/materialize/pull/21890 these tests are pretty slow:
```
  TIMEOUT [ 120.012s]                           mz-stash tests::test_stash_postgres
     PASS [ 115.726s]              mz-environmentd::auth test_auth_admin
     PASS [ 112.010s]              mz-environmentd::auth test_auth_expiry
     PASS [  77.000s]                         mz-adapter catalog::tests::test_builtin_migration
     PASS [  69.085s]                         mz-adapter catalog::tests::test_smoketest_all_builtins
```
https://buildkite.com/materialize/tests/builds/64229#018ab99a-692c-4118-88ca-9e9a3e62989e

Pinging some people who recently touched relevant cargo tests:

test_stash_postgres @mjibson 
test_auth_admin @jkosh44 
test_auth_expiry @jkosh44 @mjibson 
test_builtin_migration @jkosh44 
test_smoketest_all_builtins @jkosh44 

Getting the test faster can also decrease the total cargo test runtime, especially if the test gets scheduled towards the end.

UPDATE:
```
    REMOVED                           mz-stash tests::test_stash_postgres
    PASS [  19.617s]              mz-environmentd::auth test_auth_admin_superuser
    PASS [  19.741s]              mz-environmentd::auth test_auth_admin_non_superuser
    PASS [  56.491s]              mz-environmentd::auth test_auth_admin_show_is_superuser
    PASS [ 142.731s]              mz-environmentd::auth test_auth_expiry
    PASS [ 114.583s]                         mz-adapter catalog::tests::test_builtin_migration
    PASS [  24.842s]                         mz-adapter catalog::tests::test_smoketest_all_builtins
```

**Comment by jkosh44** [comment]
Slow bootstrapping is probably contributing to test slowness here because some of these tests start and stop Materialize multiple times.

**Comment by jkosh44** [comment]
I ran `test_builtin_migration` locally, and got a breakdown of where all the time is being spent. The test is broken up into 6 test cases, with some small amount of setup before hand. Here's an explanation of each field: 

- `Setup`: Time spent setting up the whole test.
- `stash open time`: Time spent opening the stash. This includes connecting to CRDB, creating a schema, and inserting initial values.
- `catalog open time`: Time spent opening the catalog (does not include `stash open time`). This includes, reading everything from the stash, parsing/planning/optimizing every item, and populating in-memory structs.
- `stash teardown time`: Time spent deleting the schema in CRDB.
- `Test {i}`: Total time spent running test case i, includes all stash/catalog open/close.
- `Inner test {i}`: `Test {i}` - (`stash open time` + `catalog open time` + `stash teardown time`). I.e. all the time spent actually executing the test logic outside of open/close.
- `Total time`: Time spent in the entire test.

Here's the results:

```
Setup:                  37.115µs
stash open time:        101.345134ms
catalog open time:      6.298000729s
Inner test 0:           85.01261ms
stash teardown time:    139.278027ms
Test 0:                 6.626252362s
stash open time:        96.025016ms
catalog open time:      9.531272259s
Inner test 1:           184.70547ms
stash teardown time:    283.530578ms
Test 1:                 10.101700106s
stash open time:        208.002051ms
catalog open time:      8.795265281s
Inner test 2:           179.54382ms
stash teardown time:    107.976194ms
Test 2:                 9.293081654s
stash open time:        99.915086ms
catalog open time:      6.635923376s
Inner test 3:           284.27362ms
stash teardown time:    108.628479ms
Test 3:                 7.131063765s
stash open time:        97.845881ms
catalog open time:      5.961705539s
Inner test 4:           306.062228ms
stash teardown time:    117.982639ms
Test 4:                 6.485788454s
stash open time:        104.752959ms
catalog open time:      5.969237534s
Inner test 5:           1.942393097s
stash teardown time:    124.106222ms
Test 5:                 8.142570969s
stash open time:        129.430518ms
catalog open time:      6.07035496s
Inner test 6:           177.870563ms
stash teardown time:    109.867962ms
Test 6:                 6.489697845s
Total time:             54.270227045s
```

The vast majority of time seems to be in opening the catalog which is consistent with what @mjibson has found previously. I'll work to parallelize the test and see if that helps.

**Comment by jkosh44** [comment]
Something else I have found locally, but never with this test, is sometimes deleting the CRDB schema can hang for a very long time here:

https://github.com/MaterializeInc/materialize/blob/8c82c7a9a7d1cd7ee981080abcafc463f414c62e/src/stash/src/postgres.rs#L1422-L1424

I haven't done a deep dive into it, but I've found that how my test is structured will affect if this line will hang. My guess is that if we try and drop the `DebugStashFactory` before the `Stash` itself, then the stash will be doing some background work with CRDB and block the `DROP SCHEMA` query.

**Comment by jkosh44** [comment]
Here's an actual timing of `DROP SCHEMA` (This is CRDB in a database named `materialize`).

```
materialize=> DROP SCHEMA schema_12078877448597964583 CASCADE;
ERROR:  unknown schema "schema_12078877448597964583"
Time: 584544.592 ms (09:44.545)
```

So nearly 10 minutes to ultimately do nothing. My hunch is that is CI, we run a bunch of tests in parallel, which triggers these hangs and causes the timeouts that we're seeing.

**Comment by jkosh44** [comment]
Here's an actual timing of `DROP SCHEMA` (This is CRDB in a database named `materialize`).

```
materialize=> DROP SCHEMA schema_12078877448597964583 CASCADE;
ERROR:  unknown schema "schema_12078877448597964583"
Time: 584544.592 ms (09:44.545)
```

So nearly 10 minutes to ultimately do nothing. My hunch is that is CI, we run a bunch of tests in parallel, which triggers these hangs and causes the timeouts that we're seeing.

**Comment by jkosh44** [comment]
So the problem here is the stash seems to be leaving some long running transactions open. For example below is the output of `SHOW TRANSACTIONS` in CRDB (I've added the application name locally):

```
materialize=> SHOW TRANSACTIONS;
 node_id |                txn_id                | application_name | num_stmts | num_retries | num_auto_retries | last_auto_retry_reason 
---------+--------------------------------------+------------------+-----------+-------------+------------------+------------------------
       1 | 4a2ff991-3375-4ab7-8e7e-d7b507390938 | psql             |         1 |           0 |                0 | 
       1 | 4a4837fe-54c9-4980-a19e-53cc4b877c76 | stash            |         4 |           0 |                0 | 
       1 | 95c1ef9b-b67f-43ae-97b2-065f171b414c | stash            |         0 |           0 |                0 | 
       1 | 531a5d26-c8ad-40ed-9338-5e725292e1d0 | consolidator     |         2 |           0 |                0 | 
       1 | 0b9d8fa2-7f17-4bca-8950-8220fb1842c6 | consolidator     |         0 |           0 |                0 | 
       1 | 16a57c3a-01ec-4cd9-9fed-47c664df77d7 | stash            |       136 |           0 |                0 | 
       1 | cc5e1aa2-7a58-450b-a3de-bef15b1c2d12 | stash            |        97 |           0 |                0 | 
       1 | 468b369e-827b-4134-b95a-d2f37217de73 | stash            |        44 |           0 |                0 | 
       1 | c56701fa-5606-43be-8bad-a20a4a42545d | stash            |         8 |           0 |                0 | 
(9 rows)
```

That output does not change for the duration of the test, i.e. 30 min (except for the top row which is the transaction executing `SHOW TRANSACTIONS`).

If I run the following: `SET CLUSTER SETTING sql.defaults.idle_in_transaction_session_timeout TO '5s';`, which aborts all idle transactions after 5 seconds, the test completes in 1 minute instead of 30 minutes.

It's still unclear why the stash is holding open these transactions, and why they have so many statements.

**Comment by jkosh44** [comment]
Some more interesting information, note the output of this query does change, but it happens very slowly. It will remain static for at least 1 minute at a time, usually longer.

```
SELECT t.session_id, t.start, t.application_name, t.num_stmts, t.num_retries, t.num_auto_retries, s.application_name, s.active_queries, s.last_active_query, s.status, s.num_txns_executed FROM crdb_internal.cluster_transactions t JOIN crdb_internal.cluster_sessions s ON t.session_id = s.session_id;

            session_id            |           start            |    application_name     | num_stmts | num_retries | num_auto_retries |    application_name     |                                                                                                                                                 active_queries                                                                                                                                                  |                                                                                                                                                last_active_query                                                                                                                                                | status | num_txns_executed 
----------------------------------+----------------------------+-------------------------+-----------+-------------+------------------+-------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+-------------------
 1788d67341669d180000000000000001 | 2023-09-27 20:14:22.756845 | psql                    |         1 |           0 |                0 | psql                    | SELECT t.session_id, t.start, t.application_name, t.num_stmts, t.num_retries, t.num_auto_retries, s.application_name, s.active_queries, s.last_active_query, s.status, s.num_txns_executed FROM crdb_internal.cluster_transactions AS t JOIN crdb_internal.cluster_sessions AS s ON t.session_id = s.session_id | SELECT t.session_id, t.start, t.application_name, t.num_stmts, t.num_retries, t.num_auto_retries, s.application_name, s.active_queries, s.last_active_query, s.status, s.num_txns_executed FROM crdb_internal.cluster_transactions AS t JOIN crdb_internal.cluster_sessions AS s ON t.session_id = s.session_id | ACTIVE |               121
 1788da734cad90c00000000000000001 | 2023-09-27 20:11:05.98645  | stash                   |        49 |           0 |                0 | stash                   |                                                                                                                                                                                                                                                                                                                 | SELECT key, value, "time", diff FROM data WHERE collection_id = $1                                                                                                                                                                                                                                              | IDLE   |                60
 1788da73519230360000000000000001 | 2023-09-27 20:11:05.989553 | stash                   |        43 |           0 |                0 | stash                   |                                                                                                                                                                                                                                                                                                                 | SELECT key, value, "time", diff FROM data WHERE collection_id = $1                                                                                                                                                                                                                                              | IDLE   |                54
 1788da735d431dca0000000000000001 | 2023-09-27 20:11:05.989624 | consolidator            |         2 |           0 |                0 | consolidator            |                                                                                                                                                                                                                                                                                                                 | INSERT INTO data(collection_id, key, value, "time", diff) VALUES ($1, $2, $3, $4, $5)                                                                                                                                                                                                                           | IDLE   |                38
 1788da75d826a9d90000000000000001 | 2023-09-27 20:10:59.386375 | stash                   |       215 |           0 |                0 | stash                   |                                                                                                                                                                                                                                                                                                                 | UPDATE sinces SET since = $1 WHERE collection_id = $2                                                                                                                                                                                                                                                           | IDLE   |                14
 1788da7724fc273f0000000000000001 | 2023-09-27 20:11:04.980872 | stash                   |       120 |           0 |                0 | stash                   |                                                                                                                                                                                                                                                                                                                 | SAVEPOINT txn_savepoint                                                                                                                                                                                                                                                                                         | IDLE   |                14
 1788da773771f1500000000000000001 | 2023-09-27 20:11:05.286066 | stash                   |        83 |           0 |                0 | stash                   |                                                                                                                                                                                                                                                                                                                 | INSERT INTO collections(name) VALUES ($1) RETURNING collection_id                                                                                                                                                                                                                                               | IDLE   |                14
 1788da7747ad46ef0000000000000001 | 2023-09-27 20:11:05.562374 | stash                   |        51 |           0 |                0 | stash                   |                                                                                                                                                                                                                                                                                                                 | INSERT INTO collections(name) VALUES ($1) RETURNING collection_id                                                                                                                                                                                                                                               | IDLE   |                14
 1788da813fc556df0000000000000001 | 2023-09-27 20:11:48.360652 | $ internal-create-stats |         1 |           0 |                0 | $ internal-create-stats | CREATE STATISTICS __auto__ FROM [4838] WITH OPTIONS THROTTLING 0.9 AS OF SYSTEM TIME '-30s'                                                                                                                                                                                                                     |                                                                                                                                                                                                                                                                                                                 | ACTIVE |                 1
(9 rows)
```

Some interesting things from this output:

- All the stash transaction are uncommitted and idle. That means they are sitting around not executing any statements but also not committing their transaction. What are they waiting on?
- There's a savepoint transaction open, where did that come from? I couldn't find anywhere where we open a savepoint transaction in tests.
- There's often some internal create stats transaction running. I tried running `SET CLUSTER SETTING feature.stats.enabled TO false;` which removes this transaction, but doesn't seem to help in any way.

**Comment by mjibson** [comment]
https://github.com/MaterializeInc/materialize/blob/main/src/stash/src/tests.rs#L684 opens a savepoint txn. We might need to explicitly drop the stash objects before dropping the DebugStashFactory so the connections are closed.

**Comment by mjibson** [comment]
Ugh. We already do: https://github.com/MaterializeInc/materialize/blob/main/src/stash/src/tests.rs#L719

hmmm

**Comment by jkosh44** [comment]
I had one thought, it doesn't really make sense but I'll say it anyway. Maybe when we reconnect in the stash and drop the inner client, we're not actually aborting the transaction of that client. Maybe that's what's leaving transactions open. It doesn't make sense because supposedly dropping the client should terminate the connection, but the language in the rust docs is a bit vague.

https://docs.rs/tokio-postgres/latest/tokio_postgres/struct.Connection.html

> Connection implements Future, and only resolves when the connection is closed, either because a fatal error has occurred, or because its associated Client has dropped and all outstanding work has completed.

Emphasis on "AND all outstanding work has completed".

**Comment by jkosh44** [comment]
> https://github.com/MaterializeInc/materialize/blob/d44111b700e9efaf276c2c93abd3121efa82300a/src/stash/src/tests.rs#L684
> opens a savepoint txn. We might need to explicitly drop the stash objects before dropping the DebugStashFactory so the connections are closed.

I'm only running `test_builtin_migration` and still seeing a savepoint transaction though.

**Comment by jkosh44** [comment]
> > https://github.com/MaterializeInc/materialize/blob/d44111b700e9efaf276c2c93abd3121efa82300a/src/stash/src/tests.rs#L684
> > 
> > opens a savepoint txn. We might need to explicitly drop the stash objects before dropping the DebugStashFactory so the connections are closed.
> 
> I'm only running `test_builtin_migration` and still seeing a savepoint transaction though.

I think I found the savepoint culprit:

https://github.com/MaterializeInc/materialize/blob/13a0e0442053e6d7db61b1111aacfd659d22dfbb/src/stash/src/transaction.rs#L506-L512
https://github.com/MaterializeInc/materialize/blob/13a0e0442053e6d7db61b1111aacfd659d22dfbb/src/stash/src/transaction.rs#L111-L140

I didn't realize that `append` in a stash transaction would use a savepoint.

Since the last executed statement of that transaction was `SAVEPOINT txn_savepoint` and not `RELEASE SAVEPOINT txn_savepoint` or `ROLLBACK TO SAVEPOINT txn_savepoint` maybe implies that `f().await` was taking awhile.

**Comment by mjibson** [comment]
I forgot about those savepoints! We're getting closer. That function is only called by internal stash testing (not by test_builtin_migration ) and in append. We could find out which append call by passing like a `debug_name: &str` to append and showing it somehow. Setting the application_name, appending it to txn_savepoint so it shows up in cockroach logs.

**Comment by jkosh44** [comment]
> I forgot about those savepoints! We're getting closer. That function is only called by internal stash testing (not by test_builtin_migration ) and in append. We could find out which append call by passing like a `debug_name: &str` to append and showing it somehow. Setting the application_name, appending it to txn_savepoint so it shows up in cockroach logs.

Good idea, I'll give that a shot. 

As a side note, I tried switching from schemas to database via the attached diff. Sometimes it still hangs while trying to delete the database, but sometimes it finishes in a reasonable amount of time (44 sec).

[db_change.txt](https://github.com/MaterializeInc/materialize/files/12751467/db_change.txt)


Another idea also popped into my head. We use quite a bit of mutexes in the stash. Maybe a deadlock is happening somewhere mid-transaction?

**Comment by jkosh44** [comment]
If I understand this correctly it looks like we're holding a lock while preparing a statement in CRDB, which is potentially scary.

https://github.com/MaterializeInc/materialize/blob/bda295021b71ef437f22477fe72f5b4d6ac448fa/src/stash/src/postgres.rs#L227-L251

**Comment by jkosh44** [comment]
Just for sanity sake, the following test finishes in under a second:

```
    #[mz_ore::test(tokio::test)]
    #[cfg_attr(miri, ignore)] //  unsupported operation: can't call foreign function `TLS_client_method` on OS `linux`
    async fn test_crdb_drops() {
        for _ in 0..5 {
            mz_ore::task::spawn(|| "crdb conn", async move {
                let url = std::env::var("COCKROACH_URL")
                    .expect("COCKROACH_URL environment variable is not set");
                let rng: usize = rand::thread_rng().gen();
                let schema = format!("schema_{rng}");
                let tls = mz_postgres_util::make_tls(&tokio_postgres::Config::new()).unwrap();
                let (client, connection) =
                    tokio_postgres::connect(&url, tls.clone()).await.unwrap();
                mz_ore::task::spawn(|| "tokio-postgres stash connection", async move {
                    if let Err(e) = connection.await {
                        tracing::error!("postgres stash connection error: {e}");
                    }
                });

                client
                    .batch_execute(&format!("CREATE SCHEMA {schema}"))
                    .await
                    .unwrap();
                client
                    .batch_execute(&format!("SET search_path TO {schema}"))
                    .await
                    .unwrap();
                client
                    .batch_execute("CREATE TABLE t1 (a INT)")
                    .await
                    .unwrap();
                client
                    .batch_execute("CREATE TABLE t2 (a INT)")
                    .await
                    .unwrap();
                for i in 0..20 {
                    client.batch_execute(&format!("BEGIN")).await.unwrap();
                    client
                        .batch_execute(&format!("INSERT INTO t1 VALUES ({i})"))
                        .await
                        .unwrap();
                    client
                        .batch_execute(&format!("INSERT INTO t2 SELECT a FROM t1"))
                        .await
                        .unwrap();
                    client.batch_execute(&format!("COMMIT")).await.unwrap();
                }
                client
                    .batch_execute(&format!("DROP SCHEMA {schema} CASCADE"))
                    .await
                    .unwrap();
            });
        }
    }
```

**Comment by jkosh44** [comment]
> Another idea also popped into my head. We use quite a bit of mutexes in the stash. Maybe a deadlock is happening somewhere mid-transaction?

Doesn't seem to be this. I tried removing all the locks and the issue persists.

[unlock.txt](https://github.com/MaterializeInc/materialize/files/12752142/unlock.txt)

**Comment by jkosh44** [comment]
Completely removing the consolidator also doesn't seem to resolve the issue.

[no-consolidate.txt](https://github.com/MaterializeInc/materialize/files/12752251/no-consolidate.txt)

**Comment by jkosh44** [comment]
>> I forgot about those savepoints! We're getting closer. That function is only called by internal stash testing (not by test_builtin_migration ) and in append. We could find out which append call by passing like a debug_name: &str to append and showing it somehow. Setting the application_name, appending it to txn_savepoint so it shows up in cockroach logs.
>
> Good idea, I'll give that a shot.
```
            session_id            |           start            |    application_name     | num_stmts | num_retries | num_auto_retries |    application_name     |                                                                                                                                                 active_queries                                                                                                                                                  |                                                                                                                                                last_active_query                                                                                                                                                | status | num_txns_executed 
----------------------------------+----------------------------+-------------------------+-----------+-------------+------------------+-------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+-------------------
 1789184ff84dea440000000000000001 | 2023-09-28 17:36:48.688148 | psql                    |         1 |           0 |                0 | psql                    | SELECT t.session_id, t.start, t.application_name, t.num_stmts, t.num_retries, t.num_auto_retries, s.application_name, s.active_queries, s.last_active_query, s.status, s.num_txns_executed FROM crdb_internal.cluster_transactions AS t JOIN crdb_internal.cluster_sessions AS s ON t.session_id = s.session_id | SELECT t.session_id, t.start, t.application_name, t.num_stmts, t.num_retries, t.num_auto_retries, s.application_name, s.active_queries, s.last_active_query, s.status, s.num_txns_executed FROM crdb_internal.cluster_transactions AS t JOIN crdb_internal.cluster_sessions AS s ON t.session_id = s.session_id | ACTIVE |                92
 1789208ced96d03a0000000000000001 | 2023-09-28 17:35:42.590628 | upsert-key              |         9 |           0 |                0 | upsert-key              |                                                                                                                                                                                                                                                                                                                 | UPDATE sinces SET since = $1 WHERE collection_id = $2                                                                                                                                                                                                                                                           | IDLE   |                62
 1789208cee70220a0000000000000001 | 2023-09-28 17:35:42.592808 | stash                   |         2 |           0 |                0 | stash                   |                                                                                                                                                                                                                                                                                                                 | SELECT upper FROM uppers WHERE collection_id = $1                                                                                                                                                                                                                                                               | IDLE   |                60
 1789208cfc85be480000000000000001 | 2023-09-28 17:35:42.592826 | consolidator            |         1 |           0 |                0 | consolidator            |                                                                                                                                                                                                                                                                                                                 | DELETE FROM data WHERE ((collection_id = $1) AND ("time" <= $2)) AND (key IN (SELECT key FROM data WHERE ((collection_id = $1) AND ("time" <= $2)) AND (diff < 0))) RETURNING key, value, diff                                                                                                                  | IDLE   |                46
 1789208cfd610ef70000000000000001 | 2023-09-28 17:35:42.593193 | upsert-key              |         9 |           0 |                0 | upsert-key              |                                                                                                                                                                                                                                                                                                                 | UPDATE sinces SET since = $1 WHERE collection_id = $2                                                                                                                                                                                                                                                           | IDLE   |                27
 178920912dc5d4d50000000000000001 | 2023-09-28 17:35:42.594109 | stash                   |         1 |           0 |                0 | stash                   |                                                                                                                                                                                                                                                                                                                 | SELECT EXISTS (SELECT 1 FROM pg_tables WHERE (schemaname = current_schema()) AND (tablename = 'fence'))                                                                                                                                                                                                         | IDLE   |                 5
 1789209861dbcfa40000000000000001 | 2023-09-28 17:36:13.530728 | $ internal-create-stats |         1 |           0 |                0 | $ internal-create-stats | CREATE STATISTICS __auto__ FROM [7454] WITH OPTIONS THROTTLING 0.9 AS OF SYSTEM TIME '-30s'                                                                                                                                                                                                                     |                                                                                                                                                                                                                                                                                                                 | ACTIVE |                 1
(7 rows)

```

Looks like there's a couple `upsert_key` but also a couple of transaction that didn't go through append.

Also I'm starting to distrust this output. At one point I got the following:

```
            session_id            |           start            | application_name | num_stmts | num_retries | num_auto_retries | application_name |                                                                                                                                                 active_queries                                                                                                                                                  |                                                                                                                                                last_active_query                                                                                                                                                | status | num_txns_executed 
----------------------------------+----------------------------+------------------+-----------+-------------+------------------+------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+-------------------
 1789184ff84dea440000000000000001 | 2023-09-28 15:51:48.266049 | psql             |         1 |           0 |                0 | psql             | SELECT t.session_id, t.start, t.application_name, t.num_stmts, t.num_retries, t.num_auto_retries, s.application_name, s.active_queries, s.last_active_query, s.status, s.num_txns_executed FROM crdb_internal.cluster_transactions AS t JOIN crdb_internal.cluster_sessions AS s ON t.session_id = s.session_id | SELECT t.session_id, t.start, t.application_name, t.num_stmts, t.num_retries, t.num_auto_retries, s.application_name, s.active_queries, s.last_active_query, s.status, s.num_txns_executed FROM crdb_internal.cluster_transactions AS t JOIN crdb_internal.cluster_sessions AS s ON t.session_id = s.session_id | ACTIVE |                36
 17891acdb01286400000000000000001 | 2023-09-28 15:50:23.148388 | stash            |         0 |           0 |                0 | stash            |                                                                                                                                                                                                                                                                                                                 | COMMIT TRANSACTION                                                                                                                                                                                                                                                                                              | IDLE   |                56
 17891acdb3282f8d0000000000000001 | 2023-09-28 15:50:23.139739 | stash            |        49 |           0 |                0 | stash            |                                                                                                                                                                                                                                                                                                                 | SELECT key, value, "time", diff FROM data WHERE collection_id = $1                                                                                                                                                                                                                                              | IDLE   |                47
 17891acdbb2d64450000000000000001 | 2023-09-28 15:50:23.148453 | consolidator     |         0 |           0 |                0 | consolidator     |                                                                                                                                                                                                                                                                                                                 | COMMIT TRANSACTION                                                                                                                                                                                                                                                                                              | IDLE   |                39
 17891ad1939fc3890000000000000001 | 2023-09-28 15:50:22.275425 | stash            |       129 |           0 |                0 | stash            |                                                                                                                                                                                                                                                                                                                 | RELEASE SAVEPOINT txn_savepoint                                                                                                                                                                                                                                                                                 | IDLE   |                14
 17891ad1b04989980000000000000001 | 2023-09-28 15:50:22.85696  | stash            |        44 |           0 |                0 | stash            |                                                                                                                                                                                                                                                                                                                 | SELECT key, value, "time", diff FROM data WHERE collection_id = $1                                                                                                                                                                                                                                              | IDLE   |                14
 17891ad1c03a3b860000000000000001 | 2023-09-28 15:50:23.033618 | stash            |        32 |           0 |                0 | stash            |                                                                                                                                                                                                                                                                                                                 | INSERT INTO data(collection_id, key, value, "time", diff) VALUES ($1, $2, $3, $4, $5), ($1, $6, $7, $8, $9), ($1, $10, $11, $12, $13)                                                                                                                                                                           | IDLE   |                14
 17891ad1cc0084790000000000000001 | 2023-09-28 15:50:23.142772 | stash            |         3 |           0 |                0 | stash            |                                                                                                                                                                                                                                                                                                                 | INSERT INTO collections(name) VALUES ($1) RETURNING collection_id                                                                                                                                                                                                                                               | IDLE   |                14
(8 rows)

```

How could the `last_active_query` be `COMMIT TRANSACTION` and have no `active_queries`? Wouldn't that mean that the transaction is committed and shouldn't appear here?

**Comment by jkosh44** [comment]
It doesn't seem to be related to savepoints, because it hangs even when there is no savepoint

```
            session_id            |           start            | application_name | num_stmts | num_retries | num_auto_retries | application_name |                                                                                                                                                 active_queries                                                                                                                                                  |                                                                                                                                                last_active_query                                                                                                                                                | status | num_txns_executed 
----------------------------------+----------------------------+------------------+-----------+-------------+------------------+------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+-------------------
 1789184ff84dea440000000000000001 | 2023-09-28 19:44:44.610326 | psql             |         1 |           0 |                0 | psql             | SELECT t.session_id, t.start, t.application_name, t.num_stmts, t.num_retries, t.num_auto_retries, s.application_name, s.active_queries, s.last_active_query, s.status, s.num_txns_executed FROM crdb_internal.cluster_transactions AS t JOIN crdb_internal.cluster_sessions AS s ON t.session_id = s.session_id | SELECT t.session_id, t.start, t.application_name, t.num_stmts, t.num_retries, t.num_auto_retries, s.application_name, s.active_queries, s.last_active_query, s.status, s.num_txns_executed FROM crdb_internal.cluster_transactions AS t JOIN crdb_internal.cluster_sessions AS s ON t.session_id = s.session_id | ACTIVE |               100
 1789278cf38849ac0000000000000001 | 2023-09-28 19:43:41.042833 |                  |         1 |           1 |                7 |                  | CREATE SCHEMA schema_2821601421978462189                                                                                                                                                                                                                                                                        | CREATE SCHEMA schema_2821601421978462189                                                                                                                                                                                                                                                                        | ACTIVE |                 2
 1789278d009ddcdd0000000000000001 | 2023-09-28 19:43:58.920683 | stash            |         2 |           0 |                0 | stash            |                                                                                                                                                                                                                                                                                                                 | SELECT upper FROM uppers WHERE collection_id = $1                                                                                                                                                                                                                                                               | IDLE   |                55
 1789278d017f76700000000000000001 | 2023-09-28 19:43:58.987391 | stash            |         0 |           0 |                0 | stash            |                                                                                                                                                                                                                                                                                                                 | COMMIT TRANSACTION                                                                                                                                                                                                                                                                                              | IDLE   |                51
 1789278e4c0f32240000000000000001 | 2023-09-28 19:43:58.886664 | stash            |        49 |           0 |                0 | stash            |                                                                                                                                                                                                                                                                                                                 | SELECT key, value, "time", diff FROM data WHERE collection_id = $1                                                                                                                                                                                                                                              | IDLE   |                15
(5 rows)

```

**Comment by mjibson** [comment]
Progress! I might poke at this too tomorrow.

**Comment by jkosh44** [comment]
Well I found something that fixes the issue:
[drop-diff.txt](https://github.com/MaterializeInc/materialize/files/12753639/drop-diff.txt)

I guess the shenanigans with the runtime we were doing in the `Drop` impl was somehow causing this. That doesn't really explain what CRDB was reporting in it's `crdb_internal.cluster_transactions` though.

**Comment by mjibson** [comment]
Ooh nice find. fake async drop is a footgun and apparently making a new runtime was very bad. Great!

**Comment by jkosh44** [comment]
I still have so many questions though, 

- Why were transactions being help open?
- Why did the idle transaction limit fix the issue AND still delete the schemas?
- Why doesn't the attached diff work that switches the Drop impl to a blocking API?
[blocking-diff.txt](https://github.com/MaterializeInc/materialize/files/12753772/blocking-diff.txt)


I'm not sure if it's worth it to keep digging or just push a fix and move on.

**Comment by mjibson** [comment]
https://github.com/MaterializeInc/materialize/issues/21882#issuecomment-1739801959 is an example of unexpectedly long cockroach retry timeouts bailing after 30s in production, which does feel like it's in this same bucket. If we still have those unexplained things, it does seem like this fix is a workaround instead of a full fix, and it's unfortunate that the workaround will only benefit testing.

I'm ok with shipping a know fix for now and being aware there are still concerns.

**Comment by jkosh44** [comment]
I made this branch https://github.com/jkosh44/materialize/tree/explore-migration-test-speedup to try and capture all of the CRDB queries. The logs are attached. I haven't looked through them in depth, but some of the queries are massive. Like a single query is hundreds of lines long.

[out.txt](https://github.com/MaterializeInc/materialize/files/12772842/out.txt)

**Comment by jkosh44** [comment]
Some interesting info from the new log output is that nothing happens from the time a schema drop start to the time it finishes. Also it takes roughly 5 minutes to drop the schema. 

My theory is that the call to `async_runtime.block_on` in the `Drop impl` stops all other tasks from doing work. This causes any in progress transactions (that are being run on other tasks) to become idle. The `DROP SCHEMA` sees that there are in progress transactions involving the schema and blocks until they are complete, which will never happen because the tasks will not make any progress. There must be some internal CRDB timer that waits 5 minutes and then proceeds. In fact, shortly after the schema drop is complete we see a bunch of logs that look like

```
2023-10-02T14:45:19.733856Z  INFO storage::open:storage::open_inner: mz_stash::postgres: tokio-postgres stash error, retry attempt 1: stash error: postgres: db error: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: TransactionAbortedError(ABORT_REASON_ABORT_SPAN): "sql txn" meta={id=c706b60c key=/Table/8944/1 pri=0.00864612 epo=0 ts=1696257623.117680838,0 min=1696257623.117680838,0 seq=54} lock=true stat=ABORTED rts=1696257623.117680838,0 wto=false gul=1696257623.617680838,0
HINT: See: https://www.cockroachlabs.com/docs/v22.2/transaction-retry-error-reference.html#abort_reason_abort_span, code: Some(SqlState(E40001))
```

The 5 minute timer must be how long CRDB waits before aborting the txns. This explains why setting a log idle transaction timeout fixes the issue. If I'm correct about this, then it seems to be just a testing issue and unrelated to the production timeout listed above.

**Comment by jkosh44** [comment]
Looks like the theory was correct. Adding the following before dropping schemas resolves the hanging issue:

```SQL
CANCEL SESSIONS (WITH all_s AS (SHOW SESSIONS), my_s AS (SHOW session_id) SELECT session_id FROM all_s WHERE session_id NOT IN (SELECT session_id FROM my_s))
```

**Comment by jkosh44** [comment]
I spent a decent amount of time running the auth test cases in parallel: https://github.com/MaterializeInc/materialize/pull/22121

There seems to be no difference in test length compared to the non-parallel versions.

**Comment by mjibson** [comment]
Does that imply the slowness is due to a single setup/teardown step? Did the drop stash factor change improve it?

**Comment by jkosh44** [comment]
I added some logging to the auth tests. The vast majority of time is spent in the following places:

- Starting a server. This takes roughly 8 to 15 seconds locally. A lot of the tests will start and stop a server multiple times, we should split these up into different tests so we can parallelize better.
- Waiting for token refreshes. Each one takes about 20 seconds of sitting idle doing nothing. These are calls to `wait_for_refresh`. The timeout is set in the test to 20 seconds and can easily be lowered to something faster like 1 second.

Individual test cases take less than 1 second which explains why parallelizing them doesn't help much.

I'll push these speedups later today.

**Comment by jkosh44** [comment]
Once https://github.com/MaterializeInc/materialize/pull/22138 merges the new timings will be:

```
    REMOVED                           mz-stash tests::test_stash_postgres
    PASS [  12.155s]              mz-environmentd::auth test_auth_admin_superuser
    PASS [  12.309s]              mz-environmentd::auth test_auth_admin_non_superuser
    PASS [  13.228s]              mz-environmentd::auth test_auth_admin_show_is_superuser
    PASS [  17.027s]              mz-environmentd::auth test_auth_expiry
    PASS [ 114.583s]                         mz-adapter catalog::tests::test_builtin_migration
    PASS [  24.842s]                         mz-adapter catalog::tests::test_smoketest_all_builtins
```

Which is a pretty good improvement for everything except `mz-adapter catalog::tests::test_builtin_migration`.

That one most likely suffers from slow start up times, and is testing startup specifically so it can't be easily sped up. If we still want to improve that test let's open a new issue about that test specifically.
